server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsPhase.java

```diff
@@ -56,23 +58,47 @@ public FetchSubPhaseProcessor getProcessor(FetchContext fetchContext) {
         }
 
         final SearchExecutionContext searchExecutionContext = fetchContext.getSearchExecutionContext();
-        final FieldFetcher fieldFetcher = fetchFieldsContext == null ? null
-            : fetchFieldsContext.fields() == null ? null
-            : fetchFieldsContext.fields().isEmpty() ? null
-            : FieldFetcher.create(searchExecutionContext, fetchFieldsContext.fields());
+        final FieldFetcher fieldFetcher;
+        if (fetchFieldsContext != null && fetchFieldsContext.fields() != null && fetchFieldsContext.fields().isEmpty() == false) {
+            final List<FieldAndFormat> fetchFields = fetchFieldsContext.fields()
+                .stream()
+                .filter(
+                    fieldAndFormat -> searchExecutionContext.isMetadataField(fieldAndFormat.field) == false
+                        || searchExecutionContext.getFieldType(fieldAndFormat.field).isStored() == false
+                )
+                .toList();
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, fetchFields);
+        } else {
+            fieldFetcher = FieldFetcher.create(searchExecutionContext, Collections.emptyList());
+        }
 
+        final Set<FieldAndFormat> additionalMetadataFields = new HashSet<>();
```

Comment:
This is required to:
* fetch default metadata fields **always**
* fetch metadata stored fields requested via `fields`